File: packages/Menu/fuzzy.m
#import "fuzzy.h"

FuzzyResult fuzzyMatch(NSString *query, NSString *item) {
    FuzzyResult result = {NO, 0, {0}, 0};
    NSString *lowerQuery = [query lowercaseString];
    NSString *lowerItem = [item lowercaseString];
    NSUInteger qLen = lowerQuery.length;
    NSUInteger iLen = lowerItem.length;
    if (qLen == 0) { result.match = YES; return result; }
    if (qLen > iLen) return result;

    NSUInteger qi = 0;
    int score = 0;
    BOOL prevMatched = NO;

    for (NSUInteger ii = 0; ii < iLen && qi < qLen; ii++) {
        unichar qc = [lowerQuery characterAtIndex:qi];
        unichar ic = [lowerItem characterAtIndex:ii];
        if (qc == ic) {
            result.matchIndices[result.matchCount++] = ii;
            score += 1;
            // Consecutive match bonus
            if (prevMatched) score += 3;
            // Start of string bonus
            if (ii == 0) score += 5;
            // Word boundary bonus (after space, hyphen, or camelCase)
            if (ii > 0) {
                unichar prev = [item characterAtIndex:ii - 1];
                if (prev == ' ' || prev == '-' || prev == '_')
                    score += 4;
                else if ([[NSCharacterSet uppercaseLetterCharacterSet]
                             characterIsMember:[item characterAtIndex:ii]] &&
                         [[NSCharacterSet lowercaseLetterCharacterSet]
                             characterIsMember:prev])
                    score += 3;
            }
            prevMatched = YES;
            qi++;
        } else {
            prevMatched = NO;
        }
    }

    if (qi == qLen) {
        result.match = YES;
        // Prefer shorter items
        result.score = score * 100 - (int)iLen;
    }
    return result;
}

@implementation ScoredItem
@end

File: packages/Menu/Fuzzy.swift
import Foundation

struct FuzzyResult {
    var match: Bool = false
    var score: Int = 0
    var matchIndices: [Int] = []
}

class ScoredItem {
    var text: String
    var score: Int
    var matchIndices: [Int]

    init(text: String, score: Int, matchIndices: [Int]) {
        self.text = text
        self.score = score
        self.matchIndices = matchIndices
    }
}

func fuzzyMatch(query: String, item: String) -> FuzzyResult {
    var result = FuzzyResult()
    let qChars = Array(query.lowercased())
    let iChars = Array(item.lowercased())
    let origChars = Array(item)
    let qLen = qChars.count
    let iLen = iChars.count
    if qLen == 0 { return FuzzyResult(match: true, score: 0, matchIndices: []) }
    if qLen > iLen { return result }

    var qi = 0
    var score = 0
    var prevMatched = false
    var indices: [Int] = []

    for ii in 0..<iLen {
        guard qi < qLen else { break }
        if qChars[qi] == iChars[ii] {
            indices.append(ii)
            score += 1
            if prevMatched { score += 3 }
            if ii == 0 { score += 5 }
            if ii > 0 {
                let prev = origChars[ii - 1]
                if prev == " " || prev == "-" || prev == "_" {
                    score += 4
                } else if origChars[ii].isUppercase && prev.isLowercase {
                    score += 3
                }
            }
            prevMatched = true
            qi += 1
        } else {
            prevMatched = false
        }
    }

    if qi == qLen {
        result.match = true
        result.score = score * 100 - iLen
        result.matchIndices = indices
    }
    return result
}
